go-controller/pkg/cni/cni.go

```diff
@@ -199,3 +178,23 @@ func (pr *PodRequest) getCNIResult(podInterfaceInfo *PodInterfaceInfo) (*current
 		IPs:        ips,
 	}, nil
 }
+
+func getAnnotationsOnPod(podLister corev1listers.PodLister, namespace, name string) (map[string]string, error) {
+	var annotations map[string]string
+	if err := wait.PollImmediate(1*time.Second, 30*time.Second, func() (bool, error) {
+		pod, err := podLister.Pods(namespace).Get(name)
+		if err != nil {
+			// pod is no the the cache, retryiing
```

Comment:
the problem here is if the pod is gone, we actually want to exit:

```
			if errors.IsNotFound(err) {
				// Pod not found; don't bother waiting longer
				return false, err
			}
```

otherwise we will end up waiting for a long time for no reason in the case where
1. pod add 
2. pod del
3. CNI add has started from event 1, but by the time we get here the pod is gone

Reply:
That is because you query the apiserver, this is not the same case because the pod creation is synchronous and you know that it will be there at the time you ask the apiserver, but for the cache you have to wait until the informer process the event, and that is asynchronous process and assuming that it will be there when you look can be racy.
If we have a pod deletion we should receive a CMD DEL, otherwise waiting is not a problem, we just need to wait the same time or less that CRI is waiting for our answer before giving up, right?

1. kubelet calls CRI to create a pod
2. CRI sends CMD ADD
3. cniserver needs the annotation
4. for whatever reason the pod is deleted, kubelet has to tell the CRI to send a CMD DEL

I think that the right fix is that the CMD DEL cancel the CMD ADD in that case, there is no problem in waiting, this is not blocking any other operation.

Or maybe there is a way to know if the pod was deleted in the cache ?:thinking:

Reply:
aye, pod not found we bail out :)